WalletWasabi/WabiSabi/Client/AmountDecomposer.cs

```diff
@@ -352,27 +386,97 @@ public IEnumerable<Money> Decompose(IEnumerable<Money> myInputCoinEffectiveValue
 	/// <summary>
 	/// Greedily decomposes an amount to the given denominations.
 	/// </summary>
-	private IEnumerable<ulong> BreakDown(Money coininputEffectiveValue, IEnumerable<ulong> denominations)
+	private IEnumerable<Output> BreakDown(Money coininputEffectiveValue, IEnumerable<Output> denominations)
 	{
 		var remaining = coininputEffectiveValue;
 
-		foreach (var denomPlusFee in denominations)
+		List<Output> denoms = new();
+
+		foreach (var denom in denominations)
 		{
-			if (denomPlusFee < MinAllowedOutputAmountPlusFee || remaining < MinAllowedOutputAmountPlusFee)
+			if (denom.Amount < MinAllowedOutputAmount || remaining < MinAllowedOutputAmount + ChangeFee)
 			{
 				break;
 			}
 
-			while (denomPlusFee <= remaining)
+			while (denom.EffectiveCost <= remaining)
 			{
-				yield return denomPlusFee;
-				remaining -= denomPlusFee;
+				denoms.Add(denom);
+				remaining -= denom.EffectiveCost;
 			}
 		}
 
-		if (remaining >= MinAllowedOutputAmountPlusFee)
+		if (remaining >= MinAllowedOutputAmount + ChangeFee)
 		{
-			yield return remaining;
+			denoms.Add(Output.FromAmount(remaining, ChangeScriptType, FeeRate));
 		}
+
+		return denoms;
+	}
+
+	public Money CalculateCostMetrics(IEnumerable<Output> outputs)
+	{
+		// The cost of the outputs. The more the worst.
+		var outputCost = outputs.Sum(o => o.Fee);
+
+		// The cost of sending further or remix these coins.
+		var inputCost = outputs.Sum(o => o.InputFee);
+
+		return outputCost + inputCost;
+	}
+
+	public class Output : IEqualityComparer<Output>
+	{
+		private Output(Money amount, ScriptType scriptType, FeeRate feeRate, bool isEffectiveCost)
+		{
+			ScriptType = scriptType;
+			Fee = feeRate.GetFee(scriptType.EstimateOutputVsize());
+			InputFee = feeRate.GetFee(scriptType.EstimateInputVsize());
+			// The value of amount is defined as the effective cost or a denomination amount.
+			Amount = isEffectiveCost ? amount - Fee : amount;
+		}
+
+		public Money Amount { get; }
+		public ScriptType ScriptType { get; }
+		public Money EffectiveAmount => Amount - Fee;
+		public Money EffectiveCost => Amount + Fee;
+		public Money InputFee { get; }
+		public Money Fee { get; }
+
+		public static Output FromDenomination(Money amount, ScriptType scriptType, FeeRate feeRate)
+		{
+			return new Output(amount, scriptType, feeRate, false);
+		}
+
+		public static Output FromAmount(Money amount, ScriptType scriptType, FeeRate feeRate)
+		{
+			return new Output(amount, scriptType, feeRate, true);
+		}
+
+		public bool Equals(Output? x, Output? y)
+		{
+			if (x is null || y is null)
+			{
+				if (x is null && y is null)
+				{
+					return true;
+				}
+				return false;
+			}
+
+			if (ReferenceEquals(x, y))
+			{
+				return true;
+			}
+
+			if (GetHashCode(x) == GetHashCode(y))
+			{
+				return true;
+			}
```

Comment:
This is incorrect. Two outputs with the exact same hashcode are not necessarily equal. Never use hash code to determine equality. Better to do something like this:

`return ((x.Amount, x.ScriptType, x.Fee) == (y.Amount, y.ScriptType, y.Fee)) ;`

Reply:
ACK